src/prefect/locking/filesystem.py

```diff
@@ -0,0 +1,240 @@
+                pendulum.parse(expiration) if expiration is not None else None
+            )
+            self._locks[key] = lock_info
+            return lock_info
+        except FileNotFoundError:
+            return None
+
+    def acquire_lock(
+        self,
+        key: str,
+        holder: str,
+        acquire_timeout: Optional[float] = None,
+        hold_timeout: Optional[float] = None,
+    ) -> bool:
+        self._ensure_records_directory_exists()
+        lock_path = self._lock_path_for_key(key)
+
+        if self.is_locked(key) and not self.is_lock_holder(key, holder):
+            lock_free = self.wait_for_lock(key, acquire_timeout)
+            if not lock_free:
+                return False
+
+        try:
+            Path(lock_path).touch(exist_ok=False)
+        except FileExistsError:
+            if not self.is_lock_holder(key, holder):
+                logger.debug(
+                    f"Another actor acquired the lock for record with key {key}. Trying again."
+                )
+                return self.acquire_lock(key, holder, acquire_timeout, hold_timeout)
+        expiration = (
+            pendulum.now("utc") + pendulum.duration(seconds=hold_timeout)
+            if hold_timeout is not None
+            else None
+        )
+
+        with open(Path(lock_path), "w") as lock_file:
+            json.dump(
+                {
+                    "holder": holder,
+                    "expiration": str(expiration) if expiration is not None else None,
+                },
+                lock_file,
+            )
+
+        self._locks[key] = {
+            "holder": holder,
+            "expiration": expiration,
+            "path": lock_path,
+        }
+
+        return True
+
+    async def aacquire_lock(
+        self,
+        key: str,
+        holder: str,
+        acquire_timeout: float | None = None,
+        hold_timeout: float | None = None,
+    ) -> bool:
+        await anyio.Path(self.lock_files_directory).mkdir(parents=True, exist_ok=True)
+        lock_path = self._lock_path_for_key(key)
+
+        if self.is_locked(key) and not self.is_lock_holder(key, holder):
+            lock_free = self.wait_for_lock(key, acquire_timeout)
+            if not lock_free:
+                return False
+
+        try:
+            await anyio.Path(lock_path).touch(exist_ok=False)
+        except FileExistsError:
+            if not self.is_lock_holder(key, holder):
+                logger.debug(
+                    f"Another actor acquired the lock for record with key {key}. Trying again."
+                )
+                return self.acquire_lock(key, holder, acquire_timeout, hold_timeout)
+        expiration = (
+            pendulum.now("utc") + pendulum.duration(seconds=hold_timeout)
+            if hold_timeout is not None
+            else None
+        )
+
+        async with await anyio.Path(lock_path).open("w") as lock_file:
+            await lock_file.write(
+                json.dumps(
```

Comment:
not important, but since we have `pydantic_core` already, we could use `to_json` here which is faster